Step 1 : 문자열 덧셈 계산기

JongMinCh0i opened the pull request:
리뷰 잘 부탁드립니다!

JongMinCh0i committed:
docs : 요구사항과 이를 바탕으로 한 기능 구현 목록 작성

JongMinCh0i committed:
feat : 쉼표(,) 또는 콜론(:) 을 구분자로 문자열을 분리하는 기능

JongMinCh0i committed:
feat : 문자열을 정수형으로 형 변환하는 기능 및 문자열 검증 기능

- 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw 하는 기능
- 문자열에 숫자 이외의 값이 포함되어 있는지 확인하는 기능
- 음수가 포함되어 있는 경우 RuntimeException 예외를 throw 하는 기능
- 문자열을 정수형으로 형 변환 하는 기능

JongMinCh0i committed:
feat : 문자열이 비어있는 경우("") 0을 반환하는 기능

JongMinCh0i committed:
refactor : 기능 구현 목록에 맞춘 변경

- trim 메서드 제거
- 명확한 의미의 네이밍 변경

JongMinCh0i committed:
feat : 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하고 분리한 숫자의 합을 반환하는 기능

- 문자열에 있는 숫자를 모두 더한 값을 반환하는 기능

JongMinCh0i committed:
feat : 커스텀 구분자를 지정하는 기능

- "//" "\n" 사이의 문자를 커스텀 구분자로 분리하는 기능
- 구분자의 길이가 2 이상일 경우 예외를 throw 하는 기능
- 구분자 문자열의 앞과 뒤에 "//" , "\n" 없을 경우 예외를 throw 하는 기능

JongMinCh0i committed:
fix : 명확한 테스트의 의미를 전달할 수 있는 코드 파라미터 수정

- "잘못된 커스텀 구분자 형식에 대한 예외 발생" 의 검증 의도와는 달리 커스텀 문자열의 길이에 대한 파라미터 전달로 명확한 테스트를 수행할 수 없었음으로 파라미터 제거

JongMinCh0i committed:
feat : 커스텀 구분자가 있을 경우 커스텀 구분자로 값을 더하는 기능
- 문자열이 "//" 로 시작하는지 판별하는 기능

JongMinCh0i committed:
test : 커스텀 구분자의 길이에 대한 검증 기능 테스트

JongMinCh0i committed:
refactor : 빈 문자열("")일 경우 0을 반환하는 메서드 분리를 제거하고 add 메서드 내부로 로직을 이동함

sah3122 commented on src/main/java/mission/step1/StringCalculator.java:
```diff
@@ -0,0 +1,82 @@
+
+public class StringCalculator {
+
+    public static final String DELIMITER = "[,:]";
```
상수 정의 👍 
한가지 궁금한 부분으로 Public 접근 제어자로 설정 해주신 이유가 있을까요 ?

JongMinCh0i replied:
테스트 코드에서도 프로덕션 코드에서 사용되는 상수를 사용하여 동일하게 만들려고 했었는데, 테스트 코드 수정 후 private로 변경하는 것을 누락했습니다.

혹시, 위와 같은 테스트 코드에서도 동일한 상수를 사용하고 싶은 이유라면, public 접근 제어자를 사용하는 것에 대해서 어떻게 생각하시는지 궁금합니다.

sah3122 commented on src/main/java/mission/step1/StringCalculator.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public String[] splitWithCustomDelimiter(String expression) {
+        Matcher matcher = Pattern.compile(CUSTOM_DELIMITER_PATTERN).matcher(expression);
```
매번 `Pattern.compile(CUSTOM_DELIMITER_PATTERN)` 객체를 만들지 않고 재사용할수 있는 구조는 어떻게 생각하시나요 ?

sah3122 commented on src/main/java/mission/step1/StringCalculator.java:
```diff
@@ -0,0 +1,82 @@
+            return ZERO_VALUE;
+        }
+
+        if (isCustomDelimiter(expression)) {
```
isCustomDelimiter 를 활용하여 커스텀 구분자애 대한 검증을 진행하지만 `PREFIX` 만으로 완벽한 검증이 되지 않아 splitWithCustomDelimiter 내부에서 exception을 던지고 있습니다. 
Matcher 를 기준으로 판단하는 구현도 괜찮을것 같아요 😄

sah3122 commented on src/main/java/mission/step1/StringCalculator.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    private void validatePositiveNumber(int number) {
```
음수를 보장할수 있는 클래스를 선언해보는건 어떨까요 ? 
이번 과정을 통해 최대한 작은 단위의 클래스를 도출하는 연습을 해보세요 😄

JongMinCh0i replied:
PositiveNumber 클래스를 추가하여서 양수를 보장받을 수 있게 하였습니다!

sah3122 commented on src/main/java/mission/step1/StringCalculator.java:
```diff
@@ -0,0 +1,82 @@
+    private static final String CUSTOM_DELIMITER_PATTERN = "//(.*)\n(.*)";
+    private static final String PREFIX = "//";
+
+    public String[] splitWithDelimiter(String expression) {
```
전반적으로 public method 와 private method 의 경계가 모호한것 같습니다. 
외부에 노출이 필요한 경우가 아니라면 private 접근 제어자로 막아보는건 어떨까요 ?

JongMinCh0i replied:
처음에는 계산을 담당하는 add 메서드 하나만 public 으로 두었는데, 계산을 하기 위해서는 문자열을 파싱하는 기능과, 예외처리를 하는 기능 등 서브 기능들을 테스트하기 위해서는 몇몇 기능들의 접근 제어자를 public 으로 수정할 수 밖에 없는 상황이 왔었습니다. 
혹시 이럴 때는 public 메서드인 add만 테스트를 하고 나머지 기능들은 세부적인 파라미터로 엣지케이스에 걸리는걸 눈으로 확인하는테스트 코드를 짜는것이 옳을까요? 이렇게 하게 될 경우 블랙박스 테스트가 되어서 내부 로직의 엣지 케이스를 놓칠까봐 한편으로는 어려웠던 부분입니다.

sah3122 commented on src/main/java/mission/step1/StringCalculator.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringCalculator {
```
`StringCalculator`클래스가 너무 많은 역할을 담당하고 있다고 생각해요. 
문자열을 파싱하는 클래스, 계산을 담당하는 클래스 등 적절하게 역할을 나누어 보는건 어떨까요 ?

JongMinCh0i replied:
이 부분을 수정하여 파싱하는 역할, 계산을 담당하는 역할의 클래스로 분리하였습니다. 
계산의 경우 기본 계산 , 커스텀 구분자를 활용한 계산으로 나뉘는 것을 2개의 클래스로 분리하였고, 컴포짓 패턴을 활용하여 이를 구현했습니다.

sah3122 commented on src/test/java/mission/step1/StringCalculatorTest.java:
```diff
@@ -0,0 +1,229 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+
+class StringCalculatorTest {
```
상세한 테스트 코드 작성 👍

sah3122 replied:
새롭게 추가된 클래스들에 대한 테스트 코드 작성이 필요해보입니다 😄 
이번 과정을 진행하며 단위테스트 작성에 대한 연습을 해보세요 !

sah3122 commented on src/test/java/mission/step1/StringCalculatorTest.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+
+}
```
파일의 마지막은 new line으로 끝내는것을 권장 드립니다. 
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

JongMinCh0i replied:
모든 파일의 마지막을 new line으로 끝낼 수 있게 수정하였습니다.

sah3122 reviewed (changes requested):
문자열 덧셈 계산기 구현잘해주셨습니다 👍 
몇가지 코멘트를 남겨두었는데 확인 부탁드려요 😄

JongMinCh0i committed:
refactor : 상수의 접근제어자 변경 [public -> private]

JongMinCh0i committed:
refactor : Pattern 객체를 클래스 레벨에서 한 번만 만들어서 재사용할 수 있는 구조로 변경

JongMinCh0i committed:
refactor : int 형을 래핑하는 PositiveNumber 일급 객체 생성

JongMinCh0i committed:
refactor : String 문자열을 구분자로 나누는 정책을 담당하는 책임을 별도의 인터페이스와 클래스로 분리

JongMinCh0i committed:
refactor : 계산의 정책을 담당하는 책임을 별도의 인터페이스와 클래스로 분리

JongMinCh0i committed:
feat : 커스텀 문자가 아닌 기본 입력 파라미터의 첫 번째 문자가 숫자가 아닐 경우 예외처리

JongMinCh0i committed:
refactor : 컴포짓 클래스를 활용한 전략패턴 수립

test
- 중복되는 test 코드를 제거
 - 구분자 길이가 2 이상일 때 예외 발생
- 불명확한 test 코드 파리미터 변경
 - 빈 구분자일 때 예외 발생

JongMinCh0i committed:
refactor : new Line 추가

JongMinCh0i committed:
chore : 0단계 Car 미션 패키지 이동

sah3122 commented on src/main/java/mission/step1/CalculateStrategy.java:
```diff
@@ -0,0 +1,8 @@
+package mission.step1;
+
+public interface CalculateStrategy {
```
Strategy이라고 정의해주신 이유가 있을까요 ? 
일방적인 인터페이스의 명칭이 적절할것 같습니다 😄

sah3122 commented on src/main/java/mission/step1/Calculator.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    @Override
+    public PositiveNumber calculate(PositiveNumber number) {
```
accumulated 를 상태값으로 가지지 않고 계산기는 입력 받는 값에 대한 계산만을 담당해도 될것 같아요 😄

sah3122 commented on src/main/java/mission/step1/CompositeParser.java:
```diff
@@ -0,0 +1,20 @@
+package mission.step1;
+
+public class CompositeParser implements ParserStrategy {
+    private final Parser defaultParser;
+    private final CustomParser customParser;
```
다양한 Parser 정의 잘해주셨네요 👍

sah3122 commented on src/main/java/mission/step1/PositiveNumber.java:
```diff
@@ -0,0 +1,28 @@
+package mission.step1;
+
+public class PositiveNumber {
```
양수를 보장하는 클래스 정의 👍

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
사소한 코멘트 몇가지 남겨두었는데 확인 부탁드리며 다음 미션도 화이팅 해주세요 !
